src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/token/RandomValueTokenFactory.java

```diff
@@ -0,0 +1,89 @@
+import net.oauth.OAuthMessage;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.net.URI;
+import java.util.logging.Logger;
+
+import static com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.token.ServiceProviderToken.newAccessToken;
+import static com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.token.ServiceProviderToken.newRequestToken;
+import static java.lang.System.currentTimeMillis;
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
+/**
+ * Implementation of {@link TokenFactory} that uses randomly generated values for tokens and delegates storage of
+ * generated tokens to {@link ServiceProviderTokenStore}.
+ * <br>
+ * The random token generation is based on the
+ * <a href="https://github.com/spring-projects/spring-security-oauth/blob/master/spring-security-oauth/src/main/java/org/springframework/security/oauth/provider/token/RandomValueProviderTokenServices.java">Spring OAuth Provider Library</a>
+ */
+@Singleton
+public class RandomValueTokenFactory implements TokenFactory {
+
+    private static final int ACCESS_TOKEN_SESSION_LENGTH_BYTES = 80;
+    private static final int TOKEN_SECRET_LENGTH_BYTES = 80;
+
+    private static final Logger log = Logger.getLogger(RandomValueTokenFactory.class.getName());
+
+    private final Randomizer randomizer;
+
+    @Inject
+    public RandomValueTokenFactory(Randomizer randomizer) {
+        this.randomizer = randomizer;
+    }
+
+    @Override
+    public ServiceProviderToken generateAccessToken(ServiceProviderToken requestToken) {
+        requireNonNull(requestToken, "requestToken");
+
+        if (requestToken.isAccessToken()) {
+            log.warning("Token is not a request token: " + requestToken);
+            throw new InvalidTokenException("Token is not a request token");
+        }
+
+        if (requestToken.getUser() == null || isBlank(requestToken.getVerifier())) {
+            log.warning("Request token is not authorized: " + requestToken);
+            throw new InvalidTokenException("Request token is not authorized");
+        }
+
+        log.fine(String.format("Request token '%s' was used to generate an access token", requestToken));
+
+        return newAccessToken(randomUUID().toString())
+                .callback(requestToken.getCallback())
+                .consumer(requestToken.getConsumer())
+                .creationTime(currentTimeMillis())
+                .tokenSecret(randomizer.randomUrlSafeString(TOKEN_SECRET_LENGTH_BYTES))
+                .authorizedBy(requestToken.getUser())
+                .verifier(requestToken.getVerifier())
+                .session(newSession(requestToken))
+                .build();
+    }
+
+    @Override
+    public ServiceProviderToken generateRequestToken(Consumer consumer, @Nullable URI callback, OAuthMessage message) {
```

Comment:
I am thinking if we can have 2 methods, one that takes a call back and the other which do not. We don't have to deal with nulls then. What do you think? I don't feel too strongly about this.